Remove BouncyCastle dependency from runtime

jkakavas opened the pull request:
This commit introduces a new project that builds a Jar of the
classes that have a dependency on BouncyCastle and puts it in a
subdirectory of lib (/tools/security-cli) along with the
BouncyCastle jars. This directory is then passed in the
ES_ADDITIONAL_CLASSPATH_DIRECTORIES of the CLI tools that use
these classes.

BouncyCastle is removed as a runtime dependency (remains
as a compileOnly one) from x-pack core and x-pack security.

jkakavas committed:
Remove BouncyCastle dependency from runtime

This commit introduces a new project that builds a Jar of the
classes that have a dependency on BouncyCastle and puts it in a
subdirectory of lib (/tools/security-cli) along with the
BouncyCastle jars. This directory is then passed in the
ES_ADDITIONAL_CLASSPATH_DIRECTORIES of the CLI tools that use
these classes.

BouncyCastle is removed as a runtime dependency (remains
as a compileOnly one) from x-pack core and x-pack security.

jaymode commented on distribution/build.gradle:
```diff
@@ -238,6 +238,10 @@ configure(subprojects.findAll { ['archives', 'packages'].contains(it.name) }) {
         from { project(':distribution:tools:plugin-cli').jar }
         from { project(':distribution:tools:plugin-cli').configurations.runtime }
       }
+      into('tools/security-cli') {
```
I am pretty sure this is not where we want to add this since the comment makes me think this section also applies to the oss distribution which does not have xpack.

jkakavas replied:
Thanks, that makes sense. I can take advantage of the `oss` boolean and call `with libFiles(oss)` where applicable so that it only includes the jars in `tools/security-cli` is `oss` is false

jaymode replied:
Also, I think this can probably just go under `tools` now that I see where we are putting the jar. My suggestion for this directory yesterday was due to my outdated view of how files are laid out within the distribution.

jasontedor replied:
Hmm...I think that `tools/security-cli` is the right place? It mimics what we did with the `plugin-cli`.

jaymode replied:
derp. Ignore me and leave it as is

jaymode commented on x-pack/plugin/security/cli/build.gradle:
```diff
@@ -0,0 +1,22 @@
+}
+
+jar {
+    from(project(':x-pack:plugin:core').sourceSets.main.output.classesDirs) {
```
can we not move the source files to this project? At the very least I think we should exclude these from the x-pack core jar

jkakavas replied:
Can you elaborate please? We're not moving source files to this project now. Am i reading your comment wrong or maybe the _not_ in _can we not move_ isn't meant to be there ?

I - probably falsely - assumed we do not want to move the source files to this project since we preferred this approach to the new plugin approach I tried out yesterday. It definitely makes sense to not have the classes both in this jar and in x-pack-core though - thanks.

jaymode replied:
Sorry, my comment was not clear. I'd like for us to move the source files to this project for a clearer separation.

The distinction between a separate plugin vs a jar is that the certificate generation code doesn't require a plugin and just needs to be packaged in a way that it can be loaded as a CLI tool. I suggested this way as an alternative to a completely separate plugin to avoid any confusion about seeing a `security-cli` plugin/module deployed.

jkakavas replied:
Great thanks, all clear now. I'll move the source files in this project

jasontedor commented on x-pack/plugin/security/cli/build.gradle:
```diff
@@ -0,0 +1,22 @@
+    mapping from: /bc.*/, to: 'bouncycastle'
+}
+
+licenseHeaders.enabled = false
```
Can you explain what is happening here?

jkakavas replied:
I can certainly try.  I started off without the `licenseHeaders.enabled = false` and precommit checks failed with: 

```
> Task :x-pack:plugin:security:cli:licenseHeaders FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':x-pack:plugin:security:cli:licenseHeaders'.
> You must specify at least one file to create the report for.
```

I interpreted this to be an error caused by the licenseHeaders task looking for Java files in order to check the license heades in security-cli and failing because there aren't any. Depending on whether I'll bring in the source files (https://github.com/elastic/elasticsearch/pull/32193#discussion_r203787096) , I'll remove this but what should have been the correct approach in this case?

jasontedor replied:
Sorry, I was unclear. I meant can you leave a comment in the build.gradle about this. Of course, this changes if we do bring in the source files.

jasontedor reviewed (changes requested):
I agree with @jaymode's comments and left a question.

jkakavas committed:
Address feedback

- Move source for CertificateTool, CertificateGenerateTool and
CertGenUtils and associated tests to the new project
- Ensure that the genrated jar and bc dependencies are not bundled
in oss

jkakavas commented:
@jaymode @jasontedor this is ready for another round

jaymode commented on x-pack/plugin/security/cli/build.gradle:
```diff
@@ -0,0 +1,20 @@
+    testCompile "junit:junit:${versions.junit}"
+    testCompile "org.hamcrest:hamcrest-all:${versions.hamcrest}"
+    testCompile 'org.elasticsearch:securemock:1.2'
+    testCompile("org.elasticsearch.test:framework:${version}")
```
Nit: can you change this to the same style used above for the elasticsearch compile dependency

jaymode reviewed (approved):
LGTM. I like this better and am so happy to see BC removed from the security runtime dependencies

jasontedor commented on distribution/build.gradle:
```diff
@@ -227,16 +227,24 @@ configure(subprojects.findAll { ['archives', 'packages'].contains(it.name) }) {
+          from { project(':distribution:tools:plugin-cli').jar }
+          from { project(':distribution:tools:plugin-cli').configurations.runtime }
+        }
+        if (oss == false) {
```
My only ask is a packaging test for this, that these are not in the OSS distribution, and are in the default distribution. In the new style. I am happy to guide, let me know.

jkakavas replied:
I'll take a swing at it and will ping for help if I find myself stumbling in the dark

jasontedor reviewed (approved):
Looks great! I left one comment so this is approved contingent on adding a test.

jkakavas committed:
Update class package in scripts

jkakavas committed:
address feedback

jkakavas committed:
Add packaging test for security cli

jkakavas committed:
Reintroduce compileOnly dependency for security

jkakavas committed:
ensure jimfs is available in test runtime

jkakavas committed:
Fix security-cli tests

jkakavas commented:
@jasontedor I added the a package test, please take a look and see if this resembles what you had in mind. 

Also, can you please comment on the use of `testImplementation` for the dependencies closure of security-cli. It looks like the correct one for this use case (and I do get related errors running the tests with only `testCompile`) but we are not using it anywhere else in our build scripts so I was wondering if we're solving this in some other manner.

BouncyCastle returns as a `compileOnly` dependency for security as Opensaml has a compile dependency on cryptacular which hash a compile dependency on bouncycastle.

jasontedor commented:
Thanks @jkakavas, everything looks great, including the use of `testImplementation`. Thanks a lot for adding the test.
